src/main/java/me/zeroeightsix/kami/module/modules/player/PacketLogger.kt

```diff
@@ -3,98 +3,342 @@ package me.zeroeightsix.kami.module.modules.player
+    private val packetSide by setting("Packet Side", PacketSide.BOTH, description = "Log packets from the server, from the client, or both.")
+    private val ignoreKeepAlive by setting("Ignore Keep Alive", true, description = "Ignore both incoming and outgoing KeepAlive packets.")
+    private val ignoreChunkLoading by setting("Ignore Chunk Loading", true, description = "Ignore chunk loading and unloading packets.")
+    private val ignoreUnknown by setting("Ignore Unknown Packets", false, description = "Ignore packets that aren't explicitly handled.")
+    private val ignoreChat by setting("Ignore Chat", true, description = "Ignore chat packets.")
+    private val ignoreCancelled by setting("Ignore Cancelled", true, description = "Ignore cancelled packets.")
 
-    private val file = File("${KamiMod.DIRECTORY}packet_logger.txt")
-    private val lines = Collections.synchronizedList(ArrayList<String>())
-    private val sdf = SimpleDateFormat("HH:mm:ss.SSS")
+    private val fileTimeFormatter = DateTimeFormatter.ofPattern("HH-mm-ss_SSS")
+    private val logTimeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSS")
+
+    private var start = 0L
+    private var last = 0L
+    private val timer = TickTimer(TimeUnit.SECONDS)
+
+    private var filename = ""
+    private var lines = ArrayList<String>()
+
+    private enum class PacketSide(override val displayName: String) : DisplayEnum {
+        CLIENT("Client"),
+        SERVER("Server"),
+        BOTH("Both")
+    }
 
     init {
         onEnable {
-            if (append) read()
+            start = System.currentTimeMillis()
+            filename = "${fileTimeFormatter.format(LocalTime.now())}.csv"
+
+            synchronized(this) {
+                lines.add("From,Packet Name,Time Since Start (ms),Time Since Last (ms),Data\n")
+            }
         }
 
         onDisable {
             write()
         }
 
-        safeListener<PacketEvent.Send> {
-            lines.add("${sdf.format(Date())}\n${it.packet.javaClass.simpleName}\n${it.packet.javaClass}\n${it.packet}")
-            when (it.packet) {
-                is CPacketPlayerDigging -> {
-                    lines.add("\nMining - ${it.packet.position}@${it.packet.facing} - ${it.packet.action}")
-                }
-                is CPacketPlayer -> {
-                    lines.add("\nRotation - Pitch: ${it.packet.pitch} Yaw: ${it.packet.yaw}")
+        safeListener<TickEvent.ClientTickEvent> {
+            if (it.phase != TickEvent.Phase.START) return@safeListener
+
+            if (showClientTicks) {
+                synchronized(this) {
+                    lines.add("Tick Pulse - Realtime: ${logTimeFormatter.format(LocalTime.now())} - Runtime: ${System.currentTimeMillis() - start}ms\n")
                 }
             }
-            lines.add("\n\n")
+
+            /* Don't let lines get too big, write periodically to the file */
+            if (lines.size >= 100 || timer.tick(15L)) {
```

Comment:
Maybe a bigger limit, otherwise you'll be killing your drive with how fast some packets can get